plugins/plugin-factory-github/codenvy-plugin-factory-github-server/src/main/java/com/codenvy/plugin/factory/github/server/GithubFactoryParametersResolver.java

```diff
@@ -0,0 +1,194 @@
+
+        // use master if not specified
+        if (branchName == null) {
+            branchName = "master";
+        }
+
+        // check if repository contains a codenvy Dockerfile
+        String dockerFileLocation = null;
+        String githubDockerFileLocation =
+                "https://raw.githubusercontent.com/".concat(repoUser).concat("/").concat(repoName).concat("/").concat(branchName)
+                                                    .concat("/.codenvy.dockerfile");
+        HttpURLConnection httpURLConnection = null;
+        try {
+            httpURLConnection = (HttpURLConnection)new URL(githubDockerFileLocation).openConnection();
+            httpURLConnection.setConnectTimeout(CONNECTION_TIMEOUT);
+            final int responseCode = httpURLConnection.getResponseCode();
+            if (responseCode == HttpURLConnection.HTTP_OK) {
+                dockerFileLocation = githubDockerFileLocation;
+            }
+        } catch (IOException ioe) {
+            LOG.debug("Unable to check if remote location {0} is available or not", githubDockerFileLocation);
+        } finally {
+            if (httpURLConnection != null) {
+                httpURLConnection.disconnect();
+            }
+        }
+
+        // apply default location if there is none
+        if (dockerFileLocation == null) {
+            dockerFileLocation = DEFAULT_DOCKER_FILE_LOCATION;
+        }
+
+        String envName = repoUser;
+
+        SourceStorageDto
+                sourceStorageDto = newDto(SourceStorageDto.class).withLocation(githubLocation).withType("git").withParameters(parameters);
+        ProjectConfigDto projectConfig = newDto(ProjectConfigDto.class).withSource(sourceStorageDto).withName(repoName).withType("blank")
+                                                                       .withPath("/".concat(repoName));
+
+        LimitsDto limitsDto = newDto(LimitsDto.class).withRam(2000);
+        MachineSourceDto machineSourceDto = newDto(MachineSourceDto.class).withType("dockerfile").withLocation(dockerFileLocation);
+        MachineConfigDto
+                machineConfigDto =
+                newDto(MachineConfigDto.class).withLimits(limitsDto).withType("docker").withSource(machineSourceDto).withDev(true)
+                                              .withName("ws-machine");
+
+        EnvironmentDto environmentDto = newDto(EnvironmentDto.class).withName(envName).withMachineConfigs(singletonList(machineConfigDto));
+
+        final WorkspaceConfigDto
+                workspaceConfig = newDto(WorkspaceConfigDto.class).withDefaultEnv(envName).withProjects(singletonList(projectConfig))
+                                                                  .withEnvironments(singletonList(environmentDto)).withName(repoUser);
+        final Factory factory = newDto(Factory.class).withWorkspace(workspaceConfig).withV("4.0");
+
+        return factory;
```

Comment:
Such way of constructing factory looks kinda sad for me = we strongly rely in it's model, and every change should be adopted manually. Maybe we can use something like simplified version of [this](https://github.com/codenvy/che-core/commit/a9196cc62d54928a6c2e3dbc673cec382c9bbf91#diff-d35d81a76138d9eb1ddef7605359129fL337)